doc/man/inndcomm.3
.TH INNDCOMM 3
.SH NAME
inndcomm \- INND communication part of InterNetNews library
.SH SYNOPSIS
.nf
.ta \w'    unsigned long    'u
.B
#include "inn/inndcomm.h"

.B "int"
.B "ICCopen()"

.B "int"
.B "ICCclose()"

.B "void"
.B "ICCsettimeout(i)"
.B "    int	i;"

.B "int"
.B "ICCcommand(cmd, argv, replyp)"
.B "    char	cmd;"
.B "    char	*argv[];"
.B "    char	**replyp;"

.B "int"
.B "ICCcancel(mesgid)"
.B "    char	*mesgid;"

.B "int"
.B "ICCreserve(why)"
.B "    char	*why;"

.B "int"
.B "ICCpause(why)"
.B "    char	*why;"

.B "int"
.B "ICCgo(why)"
.B "    char	*why;"

.B "extern char *ICCfailure;"
.fi
.SH DESCRIPTION
The routines described in this manual page are part of the InterNetNews
library,
.IR libinn (3).
They are used to send commands to a running
.IR innd (8)
daemon on the local host.
The letters ``ICC'' stand for
.IR I nnd
.IR C ontrol
.IR C ommand.
.PP
.I ICCopen
creates a
Unix-domain datagram socket and binds it to the server's control socket, if
.I <HAVE_UNIX_DOMAIN_SOCKETS in include/config.h>
is defined.  Otherwise it creates
a named pipe for communicating with the server.
It returns \-1 on failure or zero on success.
This routine must be called before any other routine.
.PP
.I ICCclose
closes any descriptors that have been created by
.IR ICCopen .
It returns \-1 on failure or zero on success.
.PP
.I ICCsettimeout
can be called before any of the following routines to determine how long
the library should wait before giving up on getting the server's reply.
This is done by setting and catching a SIGALRM
.IR signal (2).
If the timeout is less then zero then no reply will be waited for.
The SC_SHUTDOWN, SC_XABORT, and SC_XEXEC commands do not get a reply either.
The default, which can be obtained by setting the timeout to zero, is to
wait until the server replies.
.PP
.I ICCcommand
sends the command
.I cmd
with parameters
.I argv
to the server.
It returns \-1 on error.
If the server replies, and
.I replyp
is not NULL, it will be filled in with an allocated buffer that contains
the full text of the server's reply.
This buffer is a string in the form of ``<digits><space><text>''
where ``digits'' is the text value of the recommended exit code;
zero indicates success.
Replies longer then 4000 bytes will be truncated.
The possible values of
.I cmd
are defined in the ``inn/inndcomm.h'' header file.
The parameters for each command are described in
.IR ctlinnd (8).
This routine returns \-1 on communication failure, or the exit status
sent by the server which will never be negative.
.PP
.I ICCcancel
sends a ``cancel'' message to the server.
.I Mesgid
is the Message-ID of the article that should be cancelled.
The return value is the same as for
.IR ICCcommand .
.PP
.IR ICCpause ,
.IR ICCreserve ,
and
.I ICCgo
send a ``pause,'' ``reserve,'' or ``go'' command to the server, respectively.
If
.I ICCreserve
is used, then the
.I why
value used in the
.I ICCpause
invocation must match; the value used in the
.I ICCgo
invocation must always match that the one used in the
.I ICCpause
invocation.
The return value for all three routines is the same as for
.IR ICCcommand .
.PP
If any routine described above fails, the
.I ICCfailure
variable will identify the system call that failed.
.SH HISTORY
Written by Rich $alz <rsalz@uunet.uu.net> for InterNetNews.
.SH "SEE ALSO"
ctlinnd(8),
innd(8),
libinn(3).


NEWS
Changes in 2.4.6

    * Fixed the segfault of the radius authenticator when none of the radius
      servers respond.  Thanks to Matija Nalis for this patch.

    * Fixed a lost initialization in buffindexed, which resolves a potential
      segfault, thanks to a patch by Kirill Berezin.

    * INN now properly supports Perl 5.10.0 (and also 5.8.9); Perl filters
      were causing innd to segfault on a few systems like FreeBSD.

    * Fixed a long-standing bug which affected Perl hooks for innd: the
      variable containing the body of an article was not properly created,
      which caused regular expressions matching new lines to fail.  It
      especially affected filters like Cleanfeed which sometimes failed to
      detect unwanted articles.

      To fix that issue, Julien Elie added the use of a shared string,
      available since Perl 5.7.2, with a fall back to a slower but reliable
      copy of such bodies in case the function is not available.  Using a
      Perl version superior to 5.7.2 is therefore recommended.

    * Fixed two bugs which could prevent nnrpd from being run as a daemon in
      FreeBSD.  Thanks to Johan van Selst for having identified the problem
      and to Kai Gallasch for having provided a testing FreeBSD server.  The
      listening address was not initialized to "::0" or 0.0.0.0 when the -b
      flag was not used and an incorrect size was given when IPv6 was
      enabled and the binding done using IPv4.

    * Some annoying assertion failures occurring in innfeed have been fixed
      by Russ Allbery and Julien Elie.

    * Fixed a bug in mod-active for aliased newsgroups.  Only "=" was
      written to the active file.  Thanks to D. Stussy for this patch.

    * Fixed a bug which caused innd not to honour the Ad flag in newsfeeds.

    * Fixed a bug in the IP address displayed for "localhost" in innd's
      status file.  It was not correctly initialized.

    * Fixed a permission issue:  XHDR and XPAT were not checking the rights
      the user had to read articles when accessing them by their message-ID.

    * Fixed a bug in the replies of XHDR, XOVER and XPAT when the newsgroup
      is empty.  Two initial replies were sent instead of one:  the right
      420 code followed by a wrong 224 code.

    * When no newsgroup is selected, LISTGROUP now returns the right 412
      code (instead of 481).

    * inncheck now uses a range of permissions to see whether the file modes
      are correctly set.  Therefore, different configurations depending on
      the security the user wants to enforce on his sytem are possible.

    * A new improved version of docheckgroups is shipped with INN.  The -u
      flag permits to automatically update the newsgroups file (with a
      proper number of tabulations and an alphabetical sort), removing
      obsolete descriptions and adding new ones.  A second argument on
      command-line permits to specify which newsgroups should not be
      checked, so as not to treat them.

    * An *email=* keyword has been added by James Ralston to news.daily in
      order to supply another mail address than the one set at configure
      time for Usenet daily reports.

    * An updated moderators file with information about the aioe.*, perl.*
      and si.* hierarchies is provided; control.ctl is also up to date.

    * INN supports Berkeley DB 4.7, which is the recommended version to use
      owing to various bugs affecting previous versions of Berkeley DB.

    * Other minor bugs have also been fixed.

Changes in 2.4.5

    * Fixed the "alarm signal" around "SSL_read" in nnrpd:  it allows a
      proper disconnection of news clients which were previously hanging
      when posting an article through a SSL connection.  Moreover, the
      *clienttimeout* parameter now works on SSL connections.  Thanks to
      Matija Nalis for the patch.

    * SO_KEEPALIVE is now implemented for SSL TCP connections on systems
      which support it, allowing system detection and closing the dead TCP
      SSL connections automatically after system-specified time.  Thanks to
      Matija Nalis for the patch.

    * Fixed a segmentation fault when an article of a size greater than
      remaining stack is retrieved via SSL.  Thanks to Chris Caputo for this
      patch.

    * Fixed a few segfaults and bugs which affected both Python innd and
      nnrpd hooks.  They no longer check the existence of methods not used
      by the hooked script.  An issue with Python exception handling was
      also fixed, as well as a segfault fixed by Russ Allbery which happened
      whenever one closes and then reopens Python in the same process. 
      Julien Elie also fixed a bug when reloading Python filters (they were
      not always correctly reloaded) and a segfault when generating access
      groups with embedded Python filters for nnrpd.  Many thanks to David
      Hlacik for its bug reports.

    * The nnrpd.py stub file in order to test Python nnrpd hooks, as
      mentioned in their documentation, is now installed; only INN.py was
      previously installed in *pathfilter*.  Also fixed a bug in INN.py and
      add missing methods to it.

    * Fixed a long-standing bug in innreport which prevented it from
      correctly reporting nnrpd and innfeed log messages.

    * Fixed a hang in Perl hooks on (at least) HP/PA since Perl 5.10.

    * Fixed a compilation problem on some platforms because of AF_INET6
      which was not inside a HAVE_INET6 block in innfeed.

    * Fixed a bug in innfeed which contained thrice the same IPs for each
      peer; it unnecessarily slowed the peer IP rotation for innfeed. 
      Thanks, D. Stussy, for having seen that.  Miquel van Smoorenburg
      provided the patch.

    * A new *heavily* improved version of pullnews is shipped with this INN
      release.  This new version is provided by Geraint Edwards.  He added
      no more than 16 flags, fixed some bugs and integrated the backupfeed

lib/inndcomm.c
/*
**  Library routines to let other programs control innd.
*/

#include "config.h"
#include "clibrary.h"
#include "portable/socket.h"
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/stat.h>

#ifdef HAVE_SYS_TIME_H
# include <sys/time.h>
#endif
#include <time.h>

/* Needed on AIX 4.1 to get fd_set and friends. */
#ifdef HAVE_SYS_SELECT_H
# include <sys/select.h>
#endif

#ifdef HAVE_UNIX_DOMAIN_SOCKETS
# include "portable/socket-unix.h"
#endif

#include "inn/innconf.h"
#include "inn/inndcomm.h"
#include "inn/libinn.h"
#include "inn/paths.h"

static char			*ICCsockname = NULL;
#ifdef HAVE_UNIX_DOMAIN_SOCKETS
static struct sockaddr_un	ICCserv;
static struct sockaddr_un	ICCclient;
#endif
static int			ICCfd;
static int			ICCtimeout;
const char			*ICCfailure;


/*
**  Set the timeout.
*/
void
ICCsettimeout(int i)
{
    ICCtimeout = i;
}


/*
**  Get ready to talk to the server.
*/
int
ICCopen(void)
{
    int mask, oerrno, fd;
#ifdef HAVE_UNIX_DOMAIN_SOCKETS
    int size = 65535;
#endif

    if (innconf == NULL) {
	if (!innconf_read(NULL)) {
	    ICCfailure = "innconf";
	    return -1;
	}
    }
    /* Create a temporary name.  mkstemp is complete overkill here and is used
       only because it's convenient.  We don't use it properly, since we
       actually need to create a socket or named pipe, so there is a race
       condition here.  It doesn't matter, since pathrun isn't world-writable
       (conceivably two processes could end up using the same temporary name
       at the same time, but the worst that will happen is that one process
       will delete the other's temporary socket). */
    if (ICCsockname == NULL)
	ICCsockname = concatpath(innconf->pathrun, INN_PATH_TEMPSOCK);
    fd = mkstemp(ICCsockname);
    if (fd < 0) {
        ICCfailure = "mkstemp";
        return -1;
    }
    close(fd);
    if (unlink(ICCsockname) < 0 && errno != ENOENT) {
	ICCfailure = "unlink";
	return -1;
    }

#ifdef HAVE_UNIX_DOMAIN_SOCKETS

    /* Make a socket and give it the name. */
    if ((ICCfd = socket(AF_UNIX, SOCK_DGRAM, 0)) < 0) {
	ICCfailure = "socket";
	return -1;
    }

    /* Adjust the socket buffer size to accommodate large responses.  Ignore
       failure; the message may fit anyway, and if not, we'll fail below. */
    setsockopt(ICCfd, SOL_SOCKET, SO_RCVBUF, &size, sizeof(size));

    memset(&ICCclient, 0, sizeof ICCclient);
    ICCclient.sun_family = AF_UNIX;
    strlcpy(ICCclient.sun_path, ICCsockname, sizeof(ICCclient.sun_path));
    mask = umask(0);
    if (bind(ICCfd, (struct sockaddr *) &ICCclient, SUN_LEN(&ICCclient)) < 0) {
	oerrno = errno;
	umask(mask);
	errno = oerrno;
	ICCfailure = "bind";
	return -1;
    }
    umask(mask);

    /* Name the server's socket. */
    memset(&ICCserv, 0, sizeof ICCserv);
    ICCserv.sun_family = AF_UNIX;
    strlcpy(ICCserv.sun_path, innconf->pathrun, sizeof(ICCserv.sun_path));
    strlcat(ICCserv.sun_path, "/", sizeof(ICCserv.sun_path));
    strlcat(ICCserv.sun_path, INN_PATH_NEWSCONTROL, sizeof(ICCserv.sun_path));

#else /* !HAVE_UNIX_DOMAIN_SOCKETS */

    /* Make a named pipe and open it. */
    mask = umask(0);
    if (mkfifo(ICCsockname, 0666) < 0) {
	oerrno = errno;
	umask(mask);
	errno = oerrno;
	ICCfailure = "mkfifo";
	return -1;
    }
    umask(mask);
    if ((ICCfd = open(ICCsockname, O_RDWR)) < 0) {
	ICCfailure = "open";
	return -1;
    }
#endif /* !HAVE_UNIX_DOMAIN_SOCKETS */

    ICCfailure = NULL;
    return 0;
}


/*
**  Close down.
*/
int
ICCclose(void)
{
    int		i;

    ICCfailure = NULL;
    i = 0;
    if (close(ICCfd) < 0) {
	ICCfailure = "close";
	i = -1;
    }
    if (unlink(ICCsockname) < 0 && errno != ENOENT) {
	ICCfailure = "unlink";
	i = -1;
    }
    return i;
}


/*
**  Get the server's pid.
*/
static pid_t
ICCserverpid(void)
{
    pid_t		pid;
    FILE		*F;
    char                *path;
    char		buff[SMBUF];

    pid = 1;
    path = concatpath(innconf->pathrun, INN_PATH_SERVERPID);
    F = fopen(path, "r");
    free(path);
    if (F != NULL) {
	if (fgets(buff, sizeof buff, F) != NULL)
	    pid = atol(buff);
	fclose(F);
    }
    return pid;
}


/*
**  See if the server is still there.  When in doubt, assume yes.  Cache the
**  PID since a rebooted server won't know about our pending message.
*/
static bool
ICCserveralive(pid_t pid)
{
    if (kill(pid, 0) > 0 || errno != ESRCH)
	return true;
    return false;
}


/*
**  Send an arbitrary command to the server.
**  
**  There is a protocol version (one-byte) on the front of the message,
**  followed by a two byte length count.  The length includes the protocol
**  byte and the length itself.  This differs from the protocol in much
**  earlier versions of INN.
*/
int
ICCcommand(char cmd, const char *argv[], char **replyp)
{
    char		*buff;
    char		*p;
    const char		*q;
    char		save;
    int			i ;
#ifndef HAVE_UNIX_DOMAIN_SOCKETS
    int			fd;
    char                *path;
#endif
    int			len;
    fd_set		Rmask;
    struct timeval	T;
    pid_t		pid;
    ICC_MSGLENTYPE      rlen;
    ICC_PROTOCOLTYPE    protocol;
    size_t bufsiz = 64 * 1024 - 1;

    /* Is server there? */
    pid = ICCserverpid();
    if (!ICCserveralive(pid)) {
	ICCfailure = "dead server";
	return -1;
    }

    /* Get the length of the buffer. */
    buff = xmalloc(bufsiz);
    if (replyp)
	*replyp = NULL;

    /* Advance to leave space for length + protocol version info. */
    buff += HEADER_SIZE;
    bufsiz -= HEADER_SIZE;

    /* Format the message. */
    snprintf(buff, bufsiz, "%s%c%c", ICCsockname, SC_SEP, cmd);
    for (p = buff + strlen(buff), i = 0; (q = argv[i]) != NULL; i++) {
	*p++ = SC_SEP;
        *p = '\0';
        strlcat(buff, q, bufsiz);
        p += strlen(q);
    }

    /* Send message. */
    ICCfailure = NULL;
    len = p - buff + HEADER_SIZE;
    rlen = htons(len);

    /* now stick in the protocol version and the length. */
    buff -= HEADER_SIZE;
    bufsiz += HEADER_SIZE;
    protocol = ICC_PROTOCOL_1;
    memcpy(buff, &protocol, sizeof(protocol));
    memcpy(buff + sizeof(protocol), &rlen, sizeof(rlen));

#ifdef HAVE_UNIX_DOMAIN_SOCKETS
    if (sendto(ICCfd, buff, len, 0,(struct sockaddr *) &ICCserv,
               SUN_LEN(&ICCserv)) < 0) {
	free(buff);
	ICCfailure = "sendto";
	return -1;
    }
#else /* !HAVE_UNIX_DOMAIN_SOCKETS */
    path = concatpath(innconf->pathrun, INN_PATH_NEWSCONTROL);
    fd = open(path, O_WRONLY);
    free(path);
    if (fd < 0) {
	free(buff);
	ICCfailure = "open";
	return -1;
    }
    if (write(fd, buff, len) != len) {
	i = errno;
	free(buff);
	close(fd);
	errno = i;
	ICCfailure = "write";
	return -1;
    }
    close(fd);
#endif /* !HAVE_UNIX_DOMAIN_SOCKETS */

    /* Possibly get a reply. */
    switch (cmd) {
    default:
	if (ICCtimeout >= 0)
	    break;
	/* FALLTHROUGH */
    case SC_SHUTDOWN:
    case SC_XABORT:
    case SC_XEXEC:
	free(buff);
	return 0;
    }

    /* Wait for the reply. */
    for ( ; ; ) {
	FD_ZERO(&Rmask);
	FD_SET(ICCfd, &Rmask);
	T.tv_sec = ICCtimeout ? ICCtimeout : 120;
	T.tv_usec = 0;
	i = select(ICCfd + 1, &Rmask, NULL, NULL, &T);
	if (i < 0) {
	    free(buff);
	    ICCfailure = "select";
	    return -1;
	}
	if (i > 0 && FD_ISSET(ICCfd, &Rmask))
	    /* Server reply is there; go handle it. */
	    break;

	/* No data -- if we timed out, return. */
	if (ICCtimeout) {
	    free(buff);
	    errno = ETIMEDOUT;
	    ICCfailure = "timeout";
	    return -1;
	}

	if (!ICCserveralive(pid)) {
	    free(buff);
	    ICCfailure = "dead server";
	    return -1;
	}
    }

#ifdef HAVE_UNIX_DOMAIN_SOCKETS
    
    /* Read the reply. */
    i = RECVorREAD(ICCfd, buff, bufsiz);
    if ((unsigned int) i < HEADER_SIZE) {
        free(buff);
        ICCfailure = "read";
        return -1;
    }
    memcpy(&protocol, buff, sizeof(protocol));
    memcpy(&rlen, buff + sizeof(protocol), sizeof(rlen));
    rlen = ntohs(rlen);

    if (i != rlen) {
        free(buff);
        ICCfailure = "short read";
        return -1;
    }

    if (protocol != ICC_PROTOCOL_1) {
        free(buff);
        ICCfailure = "protocol mismatch";
        return -1;
    }

    memmove(buff, buff + HEADER_SIZE, rlen - HEADER_SIZE);
    i -= HEADER_SIZE;

    buff[i] = '\0';

#else /* !HAVE_UNIX_DOMAIN_SOCKETS */

    i = RECVorREAD(ICCfd, buff, HEADER_SIZE);
    if (i != HEADER_SIZE)
	return -1;

    memcpy(&protocol, buff, sizeof(protocol));
    memcpy(&rlen, buff + sizeof(protocol), sizeof(rlen));
    rlen = ntohs(rlen) - HEADER_SIZE;
    if (rlen > bufsiz) {
        ICCfailure = "bad length";
        return -1;
    }
    
    i = RECVorREAD(ICCfd, buff, rlen);
    if (i != rlen) {
        ICCfailure = "short read";
	return -1;
    }

    buff[i] = '\0';

    if (protocol != ICC_PROTOCOL_1) {
        ICCfailure = "protocol mismatch";
        return -1;
    }

#endif /* !HAVE_UNIX_DOMAIN_SOCKETS */
    
    /* Parse the rest of the reply; expected to be like
       <exitcode><space><text>" */
    i = 0;
    if (isdigit((unsigned char) buff[0])) {
	for (p = buff; *p && isdigit((unsigned char) *p); p++)
	    continue;
	if (*p) {
	    save = *p;
	    *p = '\0';
	    i = atoi(buff);
	    *p = save;
	}
    }
    if (replyp)
	*replyp = buff;
    else
	free(buff);

    return i;
}


/*
**  Send a "cancel" command.
*/
int
ICCcancel(const char *msgid)
{
    const char	*args[2];

    args[0] = msgid;
    args[1] = NULL;
    return ICCcommand(SC_CANCEL, args, NULL);
}


/*
**  Send a "go" command.
*/
int
ICCgo(const char *why)
{
    const char	*args[2];

    args[0] = why;
    args[1] = NULL;
    return ICCcommand(SC_GO, args, NULL);
}


/*
**  Send a "pause" command.
*/
int
ICCpause(const char *why)
{
    const char	*args[2];

    args[0] = why;
    args[1] = NULL;
    return ICCcommand(SC_PAUSE, args, NULL);
}


/*
**  Send a "reserve" command.
*/
int
ICCreserve(const char *why)
{
    const char	*args[2];

    args[0] = why;
    args[1] = NULL;
    return ICCcommand(SC_RESERVE, args, NULL);
}
